using t.Logf() as logging when testing

```console
$ go test
2022/07/21 00:41:00 [ERROR] unexpected error: unexpected error occurred
2022/07/21 00:41:00 [ERROR] unexpected error: unexpected error occurred
2022/07/21 00:41:00 [ERROR] unexpected error: unexpected error occurred
2022/07/21 00:41:00 [ERROR] unexpected error: unexpected error occurred
PASS
ok      github.com/podhmo-sandbox/sample-api/todoapp2/webapi/todo       0.014s````

--- a/experimental/define/api.go
+++ b/experimental/define/api.go
@@ -6,7 +6,6 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
-	"log"
 	"net/http"
 	"reflect"
 	"strings"
@@ -114,7 +113,7 @@ func (bc *BuildContext) BuildOpenAPIDoc(ctx context.Context) (*openapi3.T, error
 
 func (bc *BuildContext) commit(ctx context.Context) error {
 	if bc.commitFunc == nil {
-		log.Printf("[WARN]  already committed")
+		shared.GetLogger(ctx).Printf("[WARN]  already committed")
 		return nil
 	}
 	defer func() { bc.commitFunc = nil }()

--- a/experimental/validate/middleware.go
+++ b/experimental/validate/middleware.go
@@ -3,7 +3,6 @@ package validate
 import (
 	"context"
 	"fmt"
-	"log"
 	"net/http"
 	"strings"
 
@@ -96,12 +95,12 @@ func (e *Extractor) ExtractRequestValidation(ctx context.Context, req *http.Requ
 	}
 	if err := openapi3filter.ValidateRequest(ctx, input); err != nil {
 		if e.Debug {
-			log.Printf("[DEBUG] request is NG (%T) method=%s, path=%s, operationId=%s\n%+v", err, route.Method, route.Path, route.Operation.OperationID, err)
+			shared.GetLogger(ctx).Printf("[DEBUG] request is NG (%T) method=%s, path=%s, operationId=%s\n%+v", err, route.Method, route.Path, route.Operation.OperationID, err)
 		}
 		return RequestValidation{Route: &route, Input: input, Error: err}
 	}
 	if e.Debug {
-		log.Printf("[DEBUG] request is OK method=%s, path=%s, operationId=%s", route.Method, route.Path, route.Operation.OperationID)
+		shared.GetLogger(ctx).Printf("[DEBUG] request is OK method=%s, path=%s, operationId=%s", route.Method, route.Path, route.Operation.OperationID)
 	}
 	return RequestValidation{Route: &route, Input: input}
 }
@@ -121,12 +120,12 @@ func (e *Extractor) ExtractResponseValidation(ctx context.Context, validation *R
 	}
 	if err := openapi3filter.ValidateResponse(ctx, input); err != nil {
 		if e.Debug {
-			log.Printf("[DEBUG] validate response is failed: %T\n%+v", err, err)
+			shared.GetLogger(ctx).Printf("[DEBUG] validate response is failed: %T\n%+v", err, err)
 		}
 		return ResponseValidation{Route: validation.Route, Input: input, Error: err}
 	}
 	if e.Debug {
-		log.Printf("[DEBUG] response is OK") // todo: path and parameters
+		shared.GetLogger(ctx).Printf("[DEBUG] response is OK") // todo: path and parameters
 	}
 	return ResponseValidation{Route: validation.Route, Input: input}
 }

--- a/lift.go
+++ b/lift.go
@@ -19,7 +19,7 @@ func Lift[I any, O any](action Action[I, O]) http.HandlerFunc {
 		req = req.WithContext(ctx)
 
 		// binding request body and query-string and headers to input.
-		input, err := qbind.Bind[I](req, metadata)
+		input, err := qbind.Bind[I](ctx, req, metadata)
 		if err != nil {
 			code := shared.StatusCodeOfOrDefault(err, 400)
 			qdump.DumpError(w, req, err, code)

--- a/lift.go
+++ b/lift.go
@@ -15,7 +15,8 @@ type Action[I any, O any] func(ctx context.Context, input I) (output O, err erro
 func Lift[I any, O any](action Action[I, O]) http.HandlerFunc {
 	metadata := qbind.Scan(action)
 	return func(w http.ResponseWriter, req *http.Request) {
-		req = req.WithContext(shared.SetRequest(req.Context(), req)) // for shared.GetRequest() in action
+		ctx := shared.SetRequest(req.Context(), req) // for shared.GetRequest() in action
+		req = req.WithContext(ctx)
 
 		// binding request body and query-string and headers to input.
 		input, err := qbind.Bind[I](req, metadata)
@@ -28,6 +29,6 @@ func Lift[I any, O any](action Action[I, O]) http.HandlerFunc {
 		output, err := action(req.Context(), input)
 
 		// dumping result or error as json response
-		qdump.Dump(w, req, output, err)
+		qdump.Dump(ctx, w, req, output, err)
 	}
 }

--- a/qbind/bind.go
+++ b/qbind/bind.go
@@ -21,13 +21,13 @@ func init() {
 	headerDecoder.SetAliasTag("header")
 }
 
-func Bind[I any](req *http.Request, metadata Metadata) (I, error) {
+func Bind[I any](ctx context.Context, req *http.Request, metadata Metadata) (I, error) {
 	var input I
 	if metadata.HasData {
 		if req.Body == nil {
-			log.Printf("[INFO] decode json is neaded, but request body is nil, metadata=%+v, on %T", metadata, input) // TODO: structured logging
+			shared.GetLogger(ctx).Printf("[INFO] decode json is neaded, but request body is nil, metadata=%+v, on %T", metadata, input) // TODO: structured logging
 		} else if err := json.NewDecoder(req.Body).Decode(&input); err != nil {
-			log.Printf("[ERROR] unexpected error (json.Decode): %+v, on %T", err, input) // TODO: structured logging
+			shared.GetLogger(ctx).Printf("[ERROR] unexpected error (json.Decode): %+v, on %T", err, input) // TODO: structured logging
 			return input, err
 		}
 	}
@@ -40,7 +40,7 @@ func Bind[I any](req *http.Request, metadata Metadata) (I, error) {
 		}
 		if err := queryDecoder.Decode(&input, m); err != nil {
 			if shared.DEBUG {
-				log.Printf("[ERROR] unexpected query string: %+v, on %T", err, input)
+				shared.GetLogger(ctx).Printf("[ERROR] unexpected query string: %+v, on %T", err, input)
 			}
 		}
 	}
@@ -52,7 +52,7 @@ func Bind[I any](req *http.Request, metadata Metadata) (I, error) {
 		}
 		if err := headerDecoder.Decode(&input, m); err != nil {
 			if shared.DEBUG {
-				log.Printf("[ERROR] unexpected header: %+v, on %T", err, input)
+				shared.GetLogger(ctx).Printf("[ERROR] unexpected header: %+v, on %T", err, input)
 			}
 		}
 
@@ -61,7 +61,7 @@ func Bind[I any](req *http.Request, metadata Metadata) (I, error) {
 	if t, ok := any(input).(Validator); ok {
 		if err := t.Validate(req.Context()); err != nil {
 			if shared.DEBUG {
-				log.Printf("[ERROR] validation is failed: %+v, on %T", err, input)
+				shared.GetLogger(ctx).Printf("[ERROR] validation is failed: %+v, on %T", err, input)
 			}
 			return input, err
 		}

--- a/qbind/bind_test.go
+++ b/qbind/bind_test.go
@@ -8,6 +8,7 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/podhmo/or"
 	"github.com/podhmo/quickapi/qbind"
+	"github.com/podhmo/quickapi/quickapitest"
 )
 
 func TestBind(t *testing.T) {
@@ -19,9 +20,10 @@ func TestBind(t *testing.T) {
 	}
 
 	metadata := qbind.Scan(func(context.Context, input) (interface{}, error) { return nil, nil })
+	ctx := quickapitest.NewContext(t)
 	req := or.Fatal(http.NewRequest("GET", "/?pretty=true&x=y&sort=-id", nil))(t)
 
-	got := or.Fatal(qbind.Bind[input](req, metadata))(t)
+	got := or.Fatal(qbind.Bind[input](ctx, req, metadata))(t)
 	want := input{Sort: "-id", Pretty: true}
 
 	if diff := cmp.Diff(want, got); diff != "" {

--- a/qdump/dump.go
+++ b/qdump/dump.go
@@ -1,14 +1,14 @@
 package qdump
 
 import (
-	"log"
+	"context"
 	"net/http"
 
 	"github.com/go-chi/render"
 	"github.com/podhmo/quickapi/shared"
 )
 
-func Dump[O any](w http.ResponseWriter, req *http.Request, output O, err error) {
+func Dump[O any](ctx context.Context, w http.ResponseWriter, req *http.Request, output O, err error) {
 	if err != nil {
 		select {
 		case <-req.Context().Done():
@@ -25,7 +25,7 @@ func Dump[O any](w http.ResponseWriter, req *http.Request, output O, err error) 
 
 		code := shared.StatusCodeOf(err)
 		if code == 500 {
-			log.Printf("[ERROR] unexpected error: %+v", err) // TODO: structured logging
+			shared.GetLogger(ctx).Printf("[ERROR] unexpected error: %+v", err) // TODO: structured logging
 		}
 		DumpError(w, req, err, code)
 		return
@@ -36,7 +36,7 @@ func Dump[O any](w http.ResponseWriter, req *http.Request, output O, err error) 
 	}
 
 	// Force to return empty JSON array [] instead of null in case of zero slice.
-	output = FillNil(output)
+	output = FillNil(ctx, output)
 
 	render.JSON(w, req, output)
 }

--- a/qdump/dump_test.go
+++ b/qdump/dump_test.go
@@ -6,6 +6,7 @@ import (
 	"net/http/httptest"
 	"testing"
 
+	"github.com/podhmo/quickapi/quickapitest"
 	"github.com/podhmo/quickapi/shared"
 )
 
@@ -35,7 +36,8 @@ func TestConnectionIsClosed(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.msg, func(t *testing.T) {
 			rec := httptest.NewRecorder()
-			Dump[T](rec, c.request, nil, context.Canceled)
+			ctx := quickapitest.NewContext(t)
+			Dump[T](ctx, rec, c.request, nil, context.Canceled)
 
 			if want, got := c.wantCode, rec.Result().StatusCode; want != got {
 				t.Errorf("status-code in Dump(), want=%d != got=%d", want, got)
@@ -47,8 +49,9 @@ func TestConnectionIsClosed(t *testing.T) {
 func TestRedirection(t *testing.T) {
 	rec := httptest.NewRecorder()
 	req := httptest.NewRequest("GET", "/", nil)
+	ctx := quickapitest.NewContext(t)
 
-	Dump[any](rec, req, nil, shared.Redirect(http.StatusFound, "http://example.net"))
+	Dump[any](ctx, rec, req, nil, shared.Redirect(http.StatusFound, "http://example.net"))
 	res := rec.Result()
 
 	if want, got := http.StatusFound, res.StatusCode; want != got {

--- a/qdump/fill.go
+++ b/qdump/fill.go
@@ -1,27 +1,29 @@
 package qdump
 
 import (
-	"log"
+	"context"
 	"reflect"
+
+	"github.com/podhmo/quickapi/shared"
 )
 
 // TODO: performance up by qbind.Metadata
 
 // FillNil modifies the nil slice and maps it to an empty one, but this has side effects.
-func FillNil[O any](ob O) (output O) {
+func FillNil[O any](ctx context.Context, ob O) (output O) {
 	output = ob
 
 	rv := reflect.ValueOf(ob)
 	defer func() {
 		if err := recover(); err != nil {
-			log.Printf("[PANIC] unsupported kind=%s, value=%v", rv.Kind(), rv)
+			shared.GetLogger(ctx).Printf("[PANIC] unsupported kind=%s, value=%v", rv.Kind(), rv)
 		}
 	}()
 
 	if rv.Kind() == reflect.Struct {
 		rv = reflect.ValueOf(&ob).Elem() // for CanSet()
 	}
-	rv, changed := fillToplevel(rv)
+	rv, changed := fillToplevel(ctx, rv)
 	if !changed {
 		return output
 	}
@@ -32,26 +34,26 @@ var (
 	MAX_RECURSION int = 100
 )
 
-func fillToplevel(rv reflect.Value) (ret reflect.Value, changed bool) {
+func fillToplevel(ctx context.Context, rv reflect.Value) (ret reflect.Value, changed bool) {
 	switch rv.Kind() {
 	case reflect.Slice, reflect.Map:
-		if sv, changed := fill(rv, 1); changed {
+		if sv, changed := fill(ctx, rv, 1); changed {
 			return sv, true
 		}
 		return rv, false
 	case reflect.Struct:
-		return fill(rv, 1)
+		return fill(ctx, rv, 1)
 	case reflect.Pointer:
-		_, changed := fillToplevel(rv.Elem())
+		_, changed := fillToplevel(ctx, rv.Elem())
 		return rv, changed
 	default:
-		return fill(rv, 1)
+		return fill(ctx, rv, 1)
 	}
 }
 
-func fill(rv reflect.Value, lv int) (ret reflect.Value, changed bool) {
+func fill(ctx context.Context, rv reflect.Value, lv int) (ret reflect.Value, changed bool) {
 	if MAX_RECURSION <= lv {
-		log.Printf("[INFO] too deep lv=%d, kind=%s, value=%v", lv, rv.Kind(), rv)
+		shared.GetLogger(ctx).Printf("[INFO]  too deep lv=%d, kind=%s, value=%v", lv, rv.Kind(), rv)
 		return rv, false
 	}
 
@@ -66,7 +68,7 @@ func fill(rv reflect.Value, lv int) (ret reflect.Value, changed bool) {
 		case reflect.Slice, reflect.Map, reflect.Struct, reflect.Pointer: // unsafe, but for performance improvement
 			for i, n := 0, rv.Len(); i < n; i++ {
 				rf := rv.Index(i)
-				sv, subchanged := fill(rf, lv+1)
+				sv, subchanged := fill(ctx, rf, lv+1)
 				if subchanged {
 					changed = true
 					rf.Set(sv)
@@ -86,7 +88,7 @@ func fill(rv reflect.Value, lv int) (ret reflect.Value, changed bool) {
 			for iter.Next() {
 				// skip key (because: JSON's notation)
 				rf := iter.Value()
-				sv, subchanged := fill(rf, lv+1)
+				sv, subchanged := fill(ctx, rf, lv+1)
 				if subchanged {
 					changed = true
 					rv.SetMapIndex(iter.Key(), sv)
@@ -97,7 +99,7 @@ func fill(rv reflect.Value, lv int) (ret reflect.Value, changed bool) {
 	case reflect.Struct:
 		for i, n := 0, rv.NumField(); i < n; i++ {
 			rf := rv.Field(i)
-			sv, subchanged := fill(rf, lv+1)
+			sv, subchanged := fill(ctx, rf, lv+1)
 			if subchanged {
 				changed = true
 				rf.Set(sv)
@@ -117,10 +119,10 @@ func fill(rv reflect.Value, lv int) (ret reflect.Value, changed bool) {
 		return rv, false
 	case reflect.Pointer:
 		// side-effect! (not copied)
-		_, changed := fill(rv.Elem(), lv+1)
+		_, changed := fill(ctx, rv.Elem(), lv+1)
 		return rv, changed
 	default:
-		log.Printf("[ERROR] unsupported kind=%s, value=%v ...", rv.Kind(), rv)
+		shared.GetLogger(ctx).Printf("[ERROR] unsupported kind=%s, value=%v ...", rv.Kind(), rv)
 		return rv, false
 	}
 }

--- a/qdump/fill_test.go
+++ b/qdump/fill_test.go
@@ -4,6 +4,7 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+	"github.com/podhmo/quickapi/quickapitest"
 )
 
 func TestFillNil_Slice(t *testing.T) {
@@ -21,7 +22,7 @@ func TestFillNil_Slice(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.msg, func(t *testing.T) {
 			want := c.want
-			got := FillNil(c.input)
+			got := FillNil(quickapitest.NewContext(t), c.input)
 			if diff := cmp.Diff(want, got); diff != "" {
 				t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 			}
@@ -43,7 +44,7 @@ func TestFillNil_Slice2(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.msg, func(t *testing.T) {
 			want := c.want
-			got := FillNil(c.input)
+			got := FillNil(quickapitest.NewContext(t), c.input)
 			if diff := cmp.Diff(want, got); diff != "" {
 				t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 			}
@@ -55,7 +56,7 @@ func TestFillNil_Slice3(t *testing.T) {
 
 	t.Run("nil", func(t *testing.T) {
 		want := T{}
-		got := FillNil[T](nil)
+		got := FillNil[T](quickapitest.NewContext(t), nil)
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -63,7 +64,7 @@ func TestFillNil_Slice3(t *testing.T) {
 
 	t.Run("values", func(t *testing.T) {
 		want := T{{{1}, {}, {2}}, {}, {{}}}
-		got := FillNil(T{{{1}, nil, {2}}, nil, {nil}})
+		got := FillNil(quickapitest.NewContext(t), T{{{1}, nil, {2}}, nil, {nil}})
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -85,7 +86,7 @@ func TestFillNil_Map(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.msg, func(t *testing.T) {
 			want := c.want
-			got := FillNil(c.input)
+			got := FillNil(quickapitest.NewContext(t), c.input)
 			if diff := cmp.Diff(want, got); diff != "" {
 				t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 			}
@@ -110,7 +111,7 @@ func TestFillNil_Map2(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.msg, func(t *testing.T) {
 			want := c.want
-			got := FillNil(c.input)
+			got := FillNil(quickapitest.NewContext(t), c.input)
 			if diff := cmp.Diff(want, got); diff != "" {
 				t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 			}
@@ -141,7 +142,7 @@ func TestFillNil_Map3(t *testing.T) {
 	for _, c := range cases {
 		t.Run(c.msg, func(t *testing.T) {
 			want := c.want
-			got := FillNil(c.input)
+			got := FillNil(quickapitest.NewContext(t), c.input)
 			if diff := cmp.Diff(want, got); diff != "" {
 				t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 			}
@@ -157,7 +158,7 @@ func TestFillNil_Struct(t *testing.T) {
 
 	t.Run("nil", func(t *testing.T) {
 		var want *S
-		got := FillNil[*S](nil)
+		got := FillNil[*S](quickapitest.NewContext(t), nil)
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -165,7 +166,7 @@ func TestFillNil_Struct(t *testing.T) {
 
 	t.Run("self-member-nil", func(t *testing.T) {
 		want := S{Name: "Foo", Friends: []string{}}
-		got := FillNil(S{Name: "Foo"})
+		got := FillNil(quickapitest.NewContext(t), S{Name: "Foo"})
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -173,7 +174,7 @@ func TestFillNil_Struct(t *testing.T) {
 
 	t.Run("p-self-member-nil", func(t *testing.T) {
 		want := &S{Name: "Foo", Friends: []string{}}
-		got := FillNil(&S{Name: "Foo"})
+		got := FillNil(quickapitest.NewContext(t), &S{Name: "Foo"})
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -192,7 +193,7 @@ func TestFillNil_Struct_Recursive(t *testing.T) {
 
 	t.Run("nil", func(t *testing.T) {
 		var want *S
-		got := FillNil[*S](nil)
+		got := FillNil[*S](quickapitest.NewContext(t), nil)
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -200,7 +201,7 @@ func TestFillNil_Struct_Recursive(t *testing.T) {
 
 	t.Run("self", func(t *testing.T) {
 		want := S{Name: "foo", Friends: []S{}}
-		got := FillNil(S{Name: "foo"})
+		got := FillNil(quickapitest.NewContext(t), S{Name: "foo"})
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -208,7 +209,7 @@ func TestFillNil_Struct_Recursive(t *testing.T) {
 
 	t.Run("pointer-slice", func(t *testing.T) {
 		want := S{Name: "foo", Friends: []S{}, Father: &S{Friends: []S{}, Name: "father"}}
-		got := FillNil(S{Name: "foo", Father: &S{Friends: nil, Name: "father"}})
+		got := FillNil(quickapitest.NewContext(t), S{Name: "foo", Father: &S{Friends: nil, Name: "father"}})
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}
@@ -216,7 +217,7 @@ func TestFillNil_Struct_Recursive(t *testing.T) {
 
 	t.Run("pointer-slice-slice", func(t *testing.T) {
 		want := S{Name: "foo", Friends: []S{}, Father: &S{Friends: []S{{Name: "moo", Friends: []S{}}}, Name: "father"}}
-		got := FillNil(S{Name: "foo", Father: &S{Friends: []S{{Name: "moo"}}, Name: "father"}})
+		got := FillNil(quickapitest.NewContext(t), S{Name: "foo", Father: &S{Friends: []S{{Name: "moo"}}, Name: "father"}})
 		if diff := cmp.Diff(want, got); diff != "" {
 			t.Errorf("FillNil() mismatch (-want +got):\n%s", diff)
 		}

--- a/quickapitest/api.go
+++ b/quickapitest/api.go
@@ -1,12 +1,15 @@
 package quickapitest
 
 import (
+	"context"
 	"encoding/json"
 	"io"
 	"net/http"
 	"net/http/httptest"
 	"strings"
 	"testing"
+
+	"github.com/podhmo/quickapi/shared"
 )
 
 // DecodeResponse decodes json response
@@ -61,4 +64,17 @@ func DoRequest[T any](
 		opt(t, res)
 	}
 	return got
+}
+
+type TestLogger struct {
+	T *testing.T
+}
+
+func (l *TestLogger) Printf(format string, v ...any) {
+	l.T.Helper()
+	l.T.Logf(format, v...)
+}
+
+func NewContext(t *testing.T) context.Context {
+	return shared.SetLogger(context.Background(), &TestLogger{T: t})
 }

--- a/shared/context.go
+++ b/shared/context.go
@@ -2,6 +2,7 @@ package shared
 
 import (
 	"context"
+	"log"
 	"net/http"
 )
 
@@ -14,7 +15,25 @@ const (
 func GetRequest(ctx context.Context) *http.Request {
 	return ctx.Value(requestKey).(*http.Request)
 }
-
 func SetRequest(ctx context.Context, req *http.Request) context.Context {
 	return context.WithValue(ctx, requestKey, req)
+}
+
+const (
+	loggerKey contextKey = "logger"
+)
+
+type logger interface {
+	Printf(format string, v ...any)
+}
+
+func GetLogger(ctx context.Context) logger {
+	l := ctx.Value(loggerKey)
+	if l == nil {
+		return log.Default()
+	}
+	return l.(logger)
+}
+func SetLogger(ctx context.Context, l logger) context.Context {
+	return context.WithValue(ctx, loggerKey, l)
 }

